### ISSUE TITLE
useStream, useObservable

### ISSUE THREAD

**Issue by ezekielchentnik** [rehooks/ideas#19]
leveraging rxjs.  maybe we call it `useRxjs`.  could also create hooks for useXstream, useMost, etc.

[working gist](https://gist.github.com/ezekielchentnik/c924079e4b4896158b54ee470b1c2e2b)

**Comment by jamiebuilds** [comment]
That's great, maybe you can maybe it generic from RxJS by using the https://www.npmjs.com/package/symbol-observable package. I believe that's the place for the standard @benlesh?

**Comment by giuseppeg** [comment]
maybe a more generic hook for this? 
https://github.com/facebook/react/blob/master/packages/create-subscription

**Comment by ezekielchentnik** [comment]
Agreed, we can make this generic.  any suggestions on api? @benlesh

```
const [obs] = useContext(Context);
const [value, complete, next] = useObservable(obs, ['next', 'complete']);

<div>{complete ? 'complete!' : value}</div>

```

my personal use case involves connecting a redux like store using BehaviorSubject behind the scenes.  I have a Subscribe component w/render prop that injects value.

**Comment by ezekielchentnik** [comment]
I'll cook up a small implementation we can tinker with.

**Comment by skokenes** [comment]
Using observables to trigger effects would be interesting as well for situations where you need more fine-tuned control over effects being triggered than just a prop change. Something like `useEffectWhenStream`, where the second parameter is an Observable that will trigger the effect function to execute whenever the Observable fires.

A use-case that I encountered today where I used it was a side effect to post to an API whenever a prop in my component changed. However, because of the nature of the API I did not need to post the initial value of the prop - only subsequent changes. 

In a similar vein, a utility to turn prop or state changes into an Observable would be useful as well when combined with the above. Something like `useObservableProp`.

Then to solve my problem above, I did something like:
```javascript
const MyComponent = ({apiParamToPost}) => {
  // Can skip the initial value 
  const paramsToPost$ = useObservableProp(apiParamToPost).pipe(skip(1));

  // Now trigger my effect based on above
  useEffectWhenStream(() => {
    makeApiCall(apiParamToPost);
  }, paramsToPost$);

  // etc...
}
```

**Comment by fdecampredon** [comment]
I made something a bit similar : https://github.com/fdecampredon/todo-mvc-react-hooks-experiments/blob/reactive-state/src/App.js
